### PR TITLE
Fix #3303, #3305 and #3306: Implement logic to modify lesson progress

### LINKED ISSUE #3303: Implement logic for marking chapters completed
Implement logic for marking chapters completed.

**Target PR date:** 2021-06-29
**Target completion date:** 2021-07-03

### LINKED ISSUE #3305: Implement logic for marking stories completed
Implement logic for marking stories completed.

Target PR date: 2021-07-01
Target completion date: 2021-07-05

### LINKED ISSUE #3306: Implement logic for marking topics completed
Implement logic for marking topics completed.

**Target PR date:** 2021-07-03
**Target completion date:** 2021-07-07

### PR DESCRIPTION
**Pull request by yashraj-01**
## Explanation

Fixes #3303, #3305 and #3306: Implement logic to modify lesson progress.

**Scrrenshots**

- `MarkChaptersCompletedActivityTest`

| Espresso | Robolectric |
|:---:|:---:|
| Screenshot from 2021-07-08 17-53-38 | Screenshot from 2021-07-08 17-56-12 |

- `MarkStoriesCompletedActivityTest`

| Espresso | Robolectric |
|:---:|:---:|
| Screenshot from 2021-07-08 18-01-23 | Screenshot from 2021-07-08 18-02-14 |

- `MarkTopicsCompletedActivityTest`

| Espresso | Robolectric |
|:---:|:---:|
| Screenshot from 2021-07-08 18-07-59 | Screenshot from 2021-07-08 18-08-43 |

- Robolectric test results for `ModifyLessonProgressControllerTest`
Screenshot from 2021-07-08 18-10-44

## Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is made from a branch that is up-to-date with "develop".
- [ ] The PR's branch is based on "develop" and not on any other branch.
- [x] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

### CODE REVIEW

File: domain/src/main/java/org/oppia/android/domain/devoptions/ModifyLessonProgressController.kt
Comment (vinitamurthi) [inline domain/src/main/java/org/oppia/android/domain/devoptions/ModifyLessonProgressController.kt]: I am curious as to why we need to return the status boolean. And if it fails, how do we know what the issue is?

File: domain/src/main/java/org/oppia/android/domain/devoptions/ModifyLessonProgressController.kt
Comment (vinitamurthi) [inline domain/src/main/java/org/oppia/android/domain/devoptions/ModifyLessonProgressController.kt]: Same comment here and below regarding the purpose of returning the boolen

File: domain/src/test/java/org/oppia/android/domain/devoptions/ModifyLessonProgressControllerTest.kt
Comment (vinitamurthi) [inline domain/src/test/java/org/oppia/android/domain/devoptions/ModifyLessonProgressControllerTest.kt]:
```suggestion
  fun markFirstAndFractionsTopicsCompleted_all(or both)TopicsAreCompleted() {
```

File: domain/src/test/java/org/oppia/android/domain/devoptions/ModifyLessonProgressControllerTest.kt
Comment (vinitamurthi) [inline domain/src/test/java/org/oppia/android/domain/devoptions/ModifyLessonProgressControllerTest.kt]:
```suggestion
  fun markFirstAndRatios2StoriesCompleted_bothStoriesAreCompleted() {
```

File: domain/src/test/java/org/oppia/android/domain/devoptions/ModifyLessonProgressControllerTest.kt
Comment (vinitamurthi) [inline domain/src/test/java/org/oppia/android/domain/devoptions/ModifyLessonProgressControllerTest.kt]:
```suggestion
  fun markPrototypeAndBothFractionsExplorationsCompleted_allExplorationsAreCompleted() {
```

File: domain/src/main/java/org/oppia/android/domain/devoptions/ModifyLessonProgressController.kt
Comment (yashraj-01) [reply domain/src/main/java/org/oppia/android/domain/devoptions/ModifyLessonProgressController.kt]: Actually, I was using this boolean just to check if the function call was successful. You can see the use case below https://github.com/oppia/oppia-android/blob/cc53a69992335ff70106799de5f3acd2c35c4bb0/app/src/main/java/org/oppia/android/app/devoptions/marktopicscompleted/MarkTopicsCompletedFragmentPresenter.kt#L80-L86
But, I think it is unnecessary. I have removed it now. Also, I don't think that there is a possibility for this function to fail as the only reason this might fail will be because of an invalid `topicId`. But since the source of `topicId` is the list of topics fetched using https://github.com/oppia/oppia-android/blob/feb7d812b96c3b384b3ed9bfa38e3e560c058c78/domain/src/main/java/org/oppia/android/domain/devoptions/ModifyLessonProgressController.kt#L39
I don't think there will be a case of invalid `topicId`. WDYT @vinitamurthi? cc: @rt4914 @Sarthak2601.

File: domain/src/main/java/org/oppia/android/domain/devoptions/ModifyLessonProgressController.kt
Comment (yashraj-01) [reply domain/src/main/java/org/oppia/android/domain/devoptions/ModifyLessonProgressController.kt]: Explained in the previous comment. PTAL. Thanks.

File: domain/src/test/java/org/oppia/android/domain/devoptions/ModifyLessonProgressControllerTest.kt
Comment (yashraj-01) [reply domain/src/test/java/org/oppia/android/domain/devoptions/ModifyLessonProgressControllerTest.kt]: Changed to `markFirstAndFractionsTopicsCompleted_bothTopicsAreCompleted`.

File: domain/src/test/java/org/oppia/android/domain/devoptions/ModifyLessonProgressControllerTest.kt
Comment (yashraj-01) [reply domain/src/test/java/org/oppia/android/domain/devoptions/ModifyLessonProgressControllerTest.kt]: Changed to `markFirstAndRatios2StoriesCompleted_bothStoriesAreCompleted`

File: domain/src/test/java/org/oppia/android/domain/devoptions/ModifyLessonProgressControllerTest.kt
Comment (yashraj-01) [reply domain/src/test/java/org/oppia/android/domain/devoptions/ModifyLessonProgressControllerTest.kt]: Changed to `markPrototypeAndBothFractionsExplorationsCompleted_allThreeExplorationsAreCompleted`. There is some slight difference. PTAL. Thanks.

File: domain/src/main/java/org/oppia/android/domain/devoptions/ModifyLessonProgressController.kt
Comment (vinitamurthi) [reply domain/src/main/java/org/oppia/android/domain/devoptions/ModifyLessonProgressController.kt]: I think even if there is an invalid topicId, this boolean isn't really giving us a lot of information so its better to remove it.

**Review by vinitamurthi — APPROVED** [review]

File: app/src/main/java/org/oppia/android/app/devoptions/markchapterscompleted/MarkChaptersCompletedFragmentPresenter.kt
Comment (rt4914) [inline app/src/main/java/org/oppia/android/app/devoptions/markchapterscompleted/MarkChaptersCompletedFragmentPresenter.kt]: This is a bit difficult to read.
Suggestion
```
    val notCompletedChapterCount = getMarkChaptersCompletedViewModel().getItemList().count {
      it is ChapterSummaryViewModel && !it.checkIfChapterIsCompleted()
    }
    if (notCompletedChapterCount == 0) {
      this.binding.isAllChecked = true
    }
```

File: app/src/main/java/org/oppia/android/app/devoptions/markchapterscompleted/MarkChaptersCompletedViewModel.kt
Comment (rt4914) [inline app/src/main/java/org/oppia/android/app/devoptions/markchapterscompleted/MarkChaptersCompletedViewModel.kt]: Optional suggestion:
```
    itemList.filterIsInstance<ChapterSummaryViewModel>().associateBy(
      { it.chapterSummary.explorationId }, 
      { Pair(it.storyId, it.topicId) }
    )
```

File: app/src/sharedTest/java/org/oppia/android/app/devoptions/MarkChaptersCompletedActivityTest.kt
Comment (rt4914) [inline app/src/sharedTest/java/org/oppia/android/app/devoptions/MarkChaptersCompletedActivityTest.kt]: Suggest writing it as `fun testMarkChapsCompleted_land_selectChapters_clickCompleted_reopen_completedChapsAreChecked() { ` 

as in long term we will remove all `// ktlint-disable max-line-length`

File: app/src/sharedTest/java/org/oppia/android/app/devoptions/MarkChaptersCompletedActivityTest.kt
Comment (rt4914) [inline app/src/sharedTest/java/org/oppia/android/app/devoptions/MarkChaptersCompletedActivityTest.kt]: @anandwana001 Is this a good way to write tests? I meaning launching it twice in same test.

File: app/src/sharedTest/java/org/oppia/android/app/devoptions/MarkChaptersCompletedActivityTest.kt
Comment (rt4914) [inline app/src/sharedTest/java/org/oppia/android/app/devoptions/MarkChaptersCompletedActivityTest.kt]: Rename to remove `ktlint` comment.

File: app/src/sharedTest/java/org/oppia/android/app/devoptions/MarkStoriesCompletedActivityTest.kt
Comment (rt4914) [inline app/src/sharedTest/java/org/oppia/android/app/devoptions/MarkStoriesCompletedActivityTest.kt]: Ditto

File: app/src/sharedTest/java/org/oppia/android/app/devoptions/MarkTopicsCompletedActivityTest.kt
Comment (rt4914) [inline app/src/sharedTest/java/org/oppia/android/app/devoptions/MarkTopicsCompletedActivityTest.kt]: Ditto

File: app/src/sharedTest/java/org/oppia/android/app/devoptions/MarkTopicsCompletedActivityTest.kt
Comment (rt4914) [inline app/src/sharedTest/java/org/oppia/android/app/devoptions/MarkTopicsCompletedActivityTest.kt]: Ditto

**Review by rt4914 — CHANGES_REQUESTED** [review]
The `Mark Completed` text is not correctly aligned.

Also suggested some nit changes

File: app/src/sharedTest/java/org/oppia/android/app/devoptions/MarkChaptersCompletedActivityTest.kt
Comment (anandwana001) [reply app/src/sharedTest/java/org/oppia/android/app/devoptions/MarkChaptersCompletedActivityTest.kt]: not its not, make separate test cases.

File: app/src/sharedTest/java/org/oppia/android/app/devoptions/MarkChaptersCompletedActivityTest.kt
Comment (anandwana001) [inline app/src/sharedTest/java/org/oppia/android/app/devoptions/MarkChaptersCompletedActivityTest.kt]: it should be something like 

example
```
assertThat(activityTestRule.activity.isFinishing).isTrue()
```

File: app/src/sharedTest/java/org/oppia/android/app/devoptions/MarkChaptersCompletedActivityTest.kt
Comment (anandwana001) [inline app/src/sharedTest/java/org/oppia/android/app/devoptions/MarkChaptersCompletedActivityTest.kt]: ditto

File: app/src/sharedTest/java/org/oppia/android/app/devoptions/MarkChaptersCompletedActivityTest.kt
Comment (anandwana001) [reply app/src/sharedTest/java/org/oppia/android/app/devoptions/MarkChaptersCompletedActivityTest.kt]: @yashraj-01 Please check the note section - https://github.com/oppia/oppia-android/wiki/Ktlint-Guide#how-to-fix-the-most-common-issues

**Review by anandwana001 — CHANGES_REQUESTED** [review]
Left few suggestions

**Comment by yashraj-01** [comment]
> 
> 
> The `Mark Completed` text is not correctly aligned.
> 
> Also suggested some nit changes

Fixed the alignment. PTAL. Thanks.

File: app/src/main/java/org/oppia/android/app/devoptions/markchapterscompleted/MarkChaptersCompletedFragmentPresenter.kt
Comment (yashraj-01) [reply app/src/main/java/org/oppia/android/app/devoptions/markchapterscompleted/MarkChaptersCompletedFragmentPresenter.kt]: Done.

File: app/src/main/java/org/oppia/android/app/devoptions/markchapterscompleted/MarkChaptersCompletedViewModel.kt
Comment (yashraj-01) [reply app/src/main/java/org/oppia/android/app/devoptions/markchapterscompleted/MarkChaptersCompletedViewModel.kt]: Done.

File: app/src/sharedTest/java/org/oppia/android/app/devoptions/MarkChaptersCompletedActivityTest.kt
Comment (yashraj-01) [reply app/src/sharedTest/java/org/oppia/android/app/devoptions/MarkChaptersCompletedActivityTest.kt]: Done.

File: app/src/sharedTest/java/org/oppia/android/app/devoptions/MarkChaptersCompletedActivityTest.kt
Comment (yashraj-01) [reply app/src/sharedTest/java/org/oppia/android/app/devoptions/MarkChaptersCompletedActivityTest.kt]: Done.

File: app/src/sharedTest/java/org/oppia/android/app/devoptions/MarkStoriesCompletedActivityTest.kt
Comment (yashraj-01) [reply app/src/sharedTest/java/org/oppia/android/app/devoptions/MarkStoriesCompletedActivityTest.kt]: Renamed.

File: app/src/sharedTest/java/org/oppia/android/app/devoptions/MarkTopicsCompletedActivityTest.kt
Comment (yashraj-01) [reply app/src/sharedTest/java/org/oppia/android/app/devoptions/MarkTopicsCompletedActivityTest.kt]: Renamed.

File: app/src/sharedTest/java/org/oppia/android/app/devoptions/MarkTopicsCompletedActivityTest.kt
Comment (yashraj-01) [reply app/src/sharedTest/java/org/oppia/android/app/devoptions/MarkTopicsCompletedActivityTest.kt]: Renamed.

File: app/src/sharedTest/java/org/oppia/android/app/devoptions/MarkChaptersCompletedActivityTest.kt
Comment (yashraj-01) [reply app/src/sharedTest/java/org/oppia/android/app/devoptions/MarkChaptersCompletedActivityTest.kt]: Actually what I need to test is here is that when the user clicks the `MARK COMPLETED` button, the activity finishes and then when the user reopens the activity the completed items are already marked. Could you please suggest me some better ways to reopen an activity?

File: app/src/sharedTest/java/org/oppia/android/app/devoptions/MarkChaptersCompletedActivityTest.kt
Comment (yashraj-01) [reply app/src/sharedTest/java/org/oppia/android/app/devoptions/MarkChaptersCompletedActivityTest.kt]: Thanks for this. I have updated the tests. PTAL.

File: app/src/sharedTest/java/org/oppia/android/app/devoptions/MarkChaptersCompletedActivityTest.kt
Comment (yashraj-01) [reply app/src/sharedTest/java/org/oppia/android/app/devoptions/MarkChaptersCompletedActivityTest.kt]: Test updated. PTAL.

File: app/src/sharedTest/java/org/oppia/android/app/devoptions/MarkChaptersCompletedActivityTest.kt
Comment (anandwana001) [reply app/src/sharedTest/java/org/oppia/android/app/devoptions/MarkChaptersCompletedActivityTest.kt]: check this in a way
after click - as it is opening new activity - check `intended`

then for the next activity, try to pre-fill data and open the activity and test it, so this way we can have 2 separate test testing these individually, not dependent. 

Also, such test will fail on robolectric, so test one activity at a time.

Comment (yashraj-01) [reply app/src/sharedTest/java/org/oppia/android/app/devoptions/MarkChaptersCompletedActivityTest.kt]: Oh okay. In that case, I don't think we need this test as I have already written tests that check for the individual functionalities. Removing this test now. Thanks.

**Comment by yashraj-01** [comment]
@rt4914 @anandwana001 @Sarthak2601 I have updated the PR. PTAL. Thanks.

File: app/src/main/java/org/oppia/android/app/devoptions/markchapterscompleted/MarkChaptersCompletedFragmentPresenter.kt
Comment (Sarthak2601) [inline app/src/main/java/org/oppia/android/app/devoptions/markchapterscompleted/MarkChaptersCompletedFragmentPresenter.kt:90]: @rt4914 could you confirm if this is the intended functionality on click of the `Mark Completed`button?

File: domain/src/main/java/org/oppia/android/domain/devoptions/ModifyLessonProgressController.kt
Comment (Sarthak2601) [inline domain/src/main/java/org/oppia/android/domain/devoptions/ModifyLessonProgressController.kt]: Nit: `Modifies lesson progress by marking multiple topics as completed for the current user profile.  `
Ditto for other similar comments.

File: domain/src/main/java/org/oppia/android/domain/devoptions/ModifyLessonProgressController.kt
Comment (Sarthak2601) [inline domain/src/main/java/org/oppia/android/domain/devoptions/ModifyLessonProgressController.kt]: small nit: 
```suggestion
   * @param profileId: the ID corresponding to the profile for which progress needs modified.
```
ditto elsewhere.

**Review by Sarthak2601 — CHANGES_REQUESTED** [review]
Thanks @yashraj-01. Overall LGTM, suggested a few nits.

File: app/src/main/java/org/oppia/android/app/devoptions/markchapterscompleted/MarkChaptersCompletedFragmentPresenter.kt
Comment (rt4914) [reply app/src/main/java/org/oppia/android/app/devoptions/markchapterscompleted/MarkChaptersCompletedFragmentPresenter.kt:90]: Here we can either finish the activity or update it. Either one can work as this feature is meant for developers. I suggest following the solution which works easily.

Comment (yashraj-01) [reply app/src/main/java/org/oppia/android/app/devoptions/markchapterscompleted/MarkChaptersCompletedFragmentPresenter.kt:90]: Thanks for the confirmation @rt4914. I think finishing the activity is a fine approach. WDYT @Sarthak2601?

File: domain/src/main/java/org/oppia/android/domain/devoptions/ModifyLessonProgressController.kt
Comment (yashraj-01) [reply domain/src/main/java/org/oppia/android/domain/devoptions/ModifyLessonProgressController.kt]: Done.

File: domain/src/main/java/org/oppia/android/domain/devoptions/ModifyLessonProgressController.kt
Comment (yashraj-01) [reply domain/src/main/java/org/oppia/android/domain/devoptions/ModifyLessonProgressController.kt]: Done.

File: app/src/main/java/org/oppia/android/app/devoptions/markchapterscompleted/MarkChaptersCompletedFragmentPresenter.kt
Comment (Sarthak2601) [reply app/src/main/java/org/oppia/android/app/devoptions/markchapterscompleted/MarkChaptersCompletedFragmentPresenter.kt:90]: Yes, sounds good.

**Comment by yashraj-01** [comment]
Removing @BenHenning from review as he is no longer a code-owner after the merge. Thanks,

**Review by rt4914 — APPROVED** [review]
LGTM, thanks.

**Review by Sarthak2601 — APPROVED** [review]
LGTM, thanks @yashraj-01!

**Review by anandwana001 — APPROVED** [review]
LGTM

**Pull request merged**
